### PR TITLE
Code commit for P4RT-5.2 for #76

### LINKED ISSUE #76: P4RT-5.2: PacketOut

### PR DESCRIPTION
**Pull request by mitaliy**
Go script for P4RT-5.2: Traceroute: packet out
Resolves #76

### CODE REVIEW

**Comment by mitaliy** [comment]
Pass logs for testRun:

=== RUN   TestPacketOut
    traceroute_packetout_test.go:268: 
        *** Pushing topology to 10.54.147.183...
        
        
    traceroute_packetout_test.go:268: 
        *** Starting protocols on 10.54.147.183...
        
        
    traceroute_packetout_test.go:275: 
        *** Creating P4RT client for 10.54.139.164...
        
        
=== RUN   TestPacketOut/PacketOut_from_Primary_Controller
    packetout_test.go:81: Sending ipv4 packets with ttl = 0
    packetout_test.go:83: Initial number of packets: 0
    packetout_test.go:87: Sending packets now
    packetout_test.go:96: Final number of packets: 100
    packetout_test.go:99: Received 100 packets on ATE port 1/1
    packetout_test.go:81: Sending ipv4 packets with ttl = 1
    packetout_test.go:83: Initial number of packets: 100
    packetout_test.go:87: Sending packets now
    packetout_test.go:96: Final number of packets: 200
    packetout_test.go:99: Received 100 packets on ATE port 1/1
=== RUN   TestPacketOut/PacketOut_from_Primary_Controller#01
    packetout_test.go:123: Sending ipv6 packets with ttl = 0
    packetout_test.go:125: Initial number of packets: 200
    packetout_test.go:129: Sending packets now
    packetout_test.go:138: Final number of packets: 302
    packetout_test.go:141: Received 102 packets on ATE port 1/1
    packetout_test.go:123: Sending ipv6 packets with ttl = 1
    packetout_test.go:125: Initial number of packets: 302
    packetout_test.go:129: Sending packets now
    packetout_test.go:138: Final number of packets: 404
    packetout_test.go:141: Received 102 packets on ATE port 1/1
--- PASS: TestPacketOut (472.05s)
    --- PASS: TestPacketOut/PacketOut_from_Primary_Controller (164.16s)
    --- PASS: TestPacketOut/PacketOut_from_Primary_Controller#01 (160.90s)
PASS

*** Releasing the testbed...

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/traceroute_packetout_test.go [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+
+	leader := p4rt_client.P4RTClient{}
+	if err := leader.P4rtClientSet(dut.RawAPIs().P4RT(t)); err != nil {
+		t.Fatalf("Could not initialize p4rt client: %v", err)
+	}
+
+	args := &testArgs{
+		ctx:    ctx,
+		leader: &leader,
+		//follower: &follower,
```
Comment (Ankur19) [inline feature/experimental/p4rt/ate_tests/traceroute_packetout_test/traceroute_packetout_test.go:282]: Is this a TODO?

**Review by Ankur19 — CHANGES_REQUESTED** [review]

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/traceroute_packetout_test.go [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+	// Setup p4rt-client stream parameters
+	streamParameter := p4rt_client.P4RTStreamParameters{
+		Name:        streamName,
+		DeviceId:    deviceId,
+		ElectionIdH: uint64(0),
+		ElectionIdL: electionId,
+	}
+
+	// Send ClientArbitration message on both p4rt leader and follower clients.
+	clients := []*p4rt_client.P4RTClient{args.leader}
```
[the remark this answers is not in the document]
Comment (mitaliy) [reply feature/experimental/p4rt/ate_tests/traceroute_packetout_test/traceroute_packetout_test.go:205]: As per WBB Vendor Test Plan it only requires a single client for this test case (P4RT-5.2). Should I add a backup client here?

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/traceroute_packetout_test.go (hunk 1 above)
Comment (mitaliy) [reply feature/experimental/p4rt/ate_tests/traceroute_packetout_test/traceroute_packetout_test.go:282]: I’m working on removing line 282.

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+// follower client, then verify DUT interface statistics
+func testPacketOut(ctx context.Context, t *testing.T, args *testArgs) {
+	leader := args.leader
+	ttls := []int{0, 1}
+
+	packetOutTests := []struct {
+		desc       string
+		client     *p4rt_client.P4RTClient
+		expectPass bool
+	}{{
```
[the remark this answers is not in the document]
Comment (mitaliy) [reply feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go:68]: Thanks, I am working on removing that struct.

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+		desc:       "PacketOut from Primary Controller",
+		client:     leader,
+		expectPass: true,
+	}}
+
+	//for ipv4
+	for _, test := range packetOutTests {
```
[the remark this answers is not in the document]
Comment (mitaliy) [reply feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go:75]: Sure will remove it.

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+					}
+				}
+				time.Sleep(20 * time.Second)
+			}
+
+		})
+	}
+
+	//for ipv6
+	for _, test := range packetOutTests {
```
[the remark this answers is not in the document]
Comment (mitaliy) [reply feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go:117]: Sure will remove it.

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/traceroute_packetout_test.go [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+		if li == lj {
+			return idi < idj
+		}
+		return li < lj // "port2" < "port10"
+	})
+	return ports
+}
+
+// configInterfaceDUT configures the interface with the Addrs.
+func configInterfaceDUT(i *telemetry.Interface, a *attrs.Attributes) *telemetry.Interface {
```
[the remark this answers is not in the document]
Comment (mitaliy) [reply feature/experimental/p4rt/ate_tests/traceroute_packetout_test/traceroute_packetout_test.go:113]: Sure, will take care of it in my next commit.

**Comment by mitaliy** [comment]
All comments have been resolved in latest commit.

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/traceroute_packetout_test.go [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+	"github.com/cisco-open/go-p4/p4rt_client"
+	"github.com/cisco-open/go-p4/utils"
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"github.com/openconfig/featureprofiles/internal/attrs"
+	"github.com/openconfig/featureprofiles/internal/fptest"
+	"github.com/openconfig/ondatra"
+	"github.com/openconfig/ondatra/telemetry"
+	"github.com/openconfig/ygot/ygot"
+	p4_v1 "github.com/p4lang/p4runtime/go/p4/v1"
```
Comment (Ankur19) [inline feature/experimental/p4rt/ate_tests/traceroute_packetout_test/traceroute_packetout_test.go:36]: `p4_v1` to `p4v1` for consistency? https://go.dev/doc/effective_go#mixed-caps

Also all other variables using underscores

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+	desc := "PacketOut from Primary Controller"
+	ttls := []int{0, 1}
+	//for ipv4
+	t.Run(desc, func(t *testing.T) {
+		// Check initial packet counters
+		port := sortPorts(args.ate.Ports())[0].Name()
+
+		for _, ttl := range ttls {
+			t.Logf("Sending ipv4 packets with ttl = %d", ttl)
+			counter_0 := args.ate.Telemetry().Interface(port).Counters().InPkts().Get(t)
```
Comment (Ankur19) [inline feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go:71]: `counter0` https://go.dev/doc/effective_go#mixed-caps

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+			t.Logf("Initial number of packets: %d", counter_0)
+
+			packets := args.packetIO.GetPacketOut(portId, true, uint8(ttl))
+			packet_count := 100
```
Comment (Ankur19) [inline feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go:75]: `packetCounter` https://go.dev/doc/effective_go#mixed-caps

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/traceroute_packetout_test.go (hunk 7 above)
Comment (mitaliy) [reply feature/experimental/p4rt/ate_tests/traceroute_packetout_test/traceroute_packetout_test.go]: I have fixed it in next commit.

**Review by Ankur19 — APPROVED** [review]
LGTM!

I will add go reviewer. Thanks!

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/traceroute_packetout_test.go [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+
+	p2 := ate.Port(t, "port2")
+	atePort2.AddToATE(top, p2, &dutPort2)
+	return top
+}
+
+// configureDeviceId configures p4rt device-id on the DUT.
+func configureDeviceId(ctx context.Context, t *testing.T, dut *ondatra.DUTDevice) {
+	component := telemetry.Component{}
+	component.IntegratedCircuit = &telemetry.Component_IntegratedCircuit{}
```
Comment (shichuzhu) [inline feature/experimental/p4rt/ate_tests/traceroute_packetout_test/traceroute_packetout_test.go:138]: Prefer using a single struct literal to construct this.

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+func sendPackets(t *testing.T, client *p4rt_client.P4RTClient, packets []*p4v1.PacketOut, packetCount int) {
+	count := packetCount / len(packets)
+	for _, packet := range packets {
+		for i := 0; i < count; i++ {
+			if err := client.StreamChannelSendMsg(
+				&streamName, &p4v1.StreamMessageRequest{
+					Update: &p4v1.StreamMessageRequest_Packet{
+						Packet: packet,
+					},
+				}); err != nil {
```
[the remark this answers is not in the document]
Comment (mitaliy) [reply feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go:50]: Sure. I will make the change in the next commit.

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+// testPacketOut sends out PacketOut with payload on p4rt leader or
+// follower client, then verify DUT interface statistics
+func testPacketOut(ctx context.Context, t *testing.T, args *testArgs) {
+	leader := args.leader
+	desc := "PacketOut from Primary Controller"
+	ttls := []int{0, 1}
+	//for ipv4
+	t.Run(desc, func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mitaliy) [reply feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go:65]: Sure. I will make the change in the next commit.

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/traceroute_packetout_test.go (hunk 10 above)
Comment (mitaliy) [reply feature/experimental/p4rt/ate_tests/traceroute_packetout_test/traceroute_packetout_test.go:138]: Sure. I will make the change in the next commit.

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (shichuzhu) [reply feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go:65]: Ah, I was referring to the `desc` in `t.Run`

`t.Run( fmt.Sprintf ... func ...`

Sorry I'm not very familiar with review comments on github.

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/traceroute_packetout_test.go (hunk 10 above)
Comment (mitaliy) [reply feature/experimental/p4rt/ate_tests/traceroute_packetout_test/traceroute_packetout_test.go:138]: Sure, I will make this change in the following commit.

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (mitaliy) [reply feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go:65]: For the sake of consistency, I have left it as t.Logf() statements which are also displayed in the console.

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/traceroute_packetout_test.go [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+
+var (
+	p4InfoFile                                 = flag.String("p4info_file_location", "/root/ondatra/featureprofiles/feature/experimental/p4rt/wbb.p4info.pb.txt", "Path to the p4info file.")
+	p4rtNodeName                               = flag.String("p4rt_node_name", "FPC0:NPU0", "component name for P4RT Node")
+	streamName                                 = "p4rt"
+	tracerouteipv4InLayers layers.EthernetType = 0x0800
+	METADATA_INGRESS_PORT                      = uint32(1)
+	METADATA_EGRESS_PORT                       = uint32(2)
+	SUBMIT_TO_INGRESS                          = uint32(2)
+	SUBMIT_TO_EGRESS                           = uint32(0)
```
[the remark this answers is not in the document]
Comment (mitaliy) [reply feature/experimental/p4rt/ate_tests/traceroute_packetout_test/traceroute_packetout_test.go]: I have removed the variables.

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go (hunk 12 above)
Comment (shichuzhu) [inline feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go:66]: Sorry about the confusion, I meant to distinguish the desc in this and the one below for ipv6
`	t.Run(desc + " ipv4", func(t *testing.T) {`

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+
+			if counter1-counter0 < uint64(float64(packetCounter)*0.95) {
+				t.Fatalf("Not all the packets are received.")
+			}
+			time.Sleep(20 * time.Second)
+		}
+
+	})
+	//for ipv6
+	t.Run(desc, func(t *testing.T) {
```
Comment (shichuzhu) [inline feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go:100]: `	t.Run(desc + " ipv6", func(t *testing.T) {`

**Review by shichuzhu — CHANGES_REQUESTED** [review]

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go (hunk 12 above)
Comment (mitaliy) [reply feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go]: Sure. I have made the change

File: feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go (hunk 14 above)
Comment (mitaliy) [reply feature/experimental/p4rt/ate_tests/traceroute_packetout_test/packetout_test.go]: Sure. I have made the change in the next commit

**Comment by mitaliy** [comment]
All comments have been addressed in the next commit

**Comment by Ankur19** [comment]
@mitaliy can you also rebase to `featureprofiles/main`. This will fix the featureprofiles-pr check.

**Review by Ankur19 — APPROVED** [review]

**Review by shichuzhu — APPROVED** [review]

**Pull request merged**